Interactivity API: Use modules instead of scripts in the frontend

luisherranz opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Built on top of https://github.com/WordPress/gutenberg/pull/56118.

This is the third of a series of experiments to understand the requirements of a hypothetical Modules API. More info in the Tracking Issue:

- https://github.com/WordPress/gutenberg/issues/55942

This experiment aims to implement an API that is more complex than the ones implemented in the [first](https://github.com/WordPress/gutenberg/pull/56092) and [second](https://github.com/WordPress/gutenberg/pull/56118) experiments, adding a required server dependency graph to be able to print only the necessary modules in the import map, apart from preloading modules to avoid initial waterfalls:

- An API to register modules using:
  - A module identifier (required)
  - The full URL or the path (required)
  - The static and dynamic dependencies (required)
  - A version (optional)
- An API to enqueue modules.
- Only the modules needed for the enqueued modules have an entry in the import map.
- Enqueued modules add a `<script>` tag in the head.
- The static dependencies are preloaded when the module is enqueued using a `<link rel="modulepreload">` tag in the head.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

In the [first experiment](https://github.com/WordPress/gutenberg/pull/56092), it's not easy to know which modules need to be preloaded and, therefore, it's not easy to avoid the initial waterfalls.

In the [second experiment](https://github.com/WordPress/gutenberg/pull/56118), all the registered modules are added to the import map, even the ones that are not used on the page. In this version, declaring the dependencies is required, so we can know all the modules that are going to be used on the page and only include those in the import map.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

By leveraging a new `Gutenberg_Modules` class that acts as a singleton and a couple of functions to expose the functionality:

- `gutenberg_register_module` to register modules and add entries to the import map.
- `gutenberg_enqueue_module` to enqueue modules.

Then, we print three times using the `wp_head` action:

- The import map.
- The enqueued modules.
- The preloaded modules.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

- Add some interactive blocks: Navigation, Query or Image block.
- Check that in the frontend, there is an import map script.
- Check that scripts with `type="module"` are loaded by the interactive blocks.
- Check that the `@wordpress/interactivity` module is loaded by the other modules using `import ... from "@wordpress/interactivity"`.
- Check that the `@wordpress/interactivity` module is downloaded at the same time as the rest (instead of after, like in the [first experiment](https://github.com/WordPress/gutenberg/pull/56092)), and that there is a link tag with `rel="modulepreload"` in the head.
- Check that the import map only has the `@wordpress/interactivity` module.


## Follow-up tasks

- [ ] Replace the syntax of the dependencies array with [this one](https://github.com/WordPress/gutenberg/pull/56143#discussion_r1395753721).
- [ ] Filter enqueued modules when printing the preloaded modules.

luisherranz committed:
Bundle `@wordpress/interactivity` as an ES module

luisherranz committed:
Add the class with a basic API

luisherranz committed:
Make it work with the Navigation block

luisherranz committed:
Add versions

luisherranz committed:
Register `@wordpress/interactivity` module

luisherranz committed:
Add query and image blocks

luisherranz committed:
Add id with module identifier to the enqueued modules

luisherranz committed:
Add requried $usage for admin/frontend

luisherranz committed:
Avoid the use of enqueue to register modules

luisherranz committed:
Refactor versions

luisherranz committed:
Switch from `both` to `['admin', 'frontend']`

luisherranz committed:
Improve comments

luisherranz committed:
Add an optional static dependency graph

luisherranz committed:
Add static and dynamic dependencies in the server

luisherranz committed:
Add the file and search blocks

luisherranz committed:
Move $version out of $args to match wp_register_script

luisherranz committed:
Improve version logic

youknowriad commented on lib/experimental/modules/class-gutenberg-modules.php:
```diff
@@ -0,0 +1,186 @@
+	 *
+	 * @param string           $module_identifier The identifier of the module. Should be unique. It will be used in the final import map.
+	 * @param string           $src               Full URL of the module, or path of the script relative to the WordPress root directory.
+	 * @param array            $dependencies      Optional. An array of module identifiers of the static and dynamic dependencies of this module. It can be an indexed array, in which case all the dependencies are static, or it can be an associative array, in which case it has to contain the keys `static` and `dynamic`.
```
I wonder if instead of having subkey here  to separate these two into two top level arrays (though harder to name them)

youknowriad replied:
I guess my question also relates to the fact that if we ever want to move from "server graph" to "no server graph" at some point, the API would stay the same (no breaking change), we'd just change how we interpret it.

luisherranz replied:
So, you mean something like this, right?

```
wp_register_module(
  - string $module_identifier,
  - string|false $src,
  - string[] $static_deps = array(),
  - string[] $dynamic_deps = array(),
  - string|bool|null $ver = false
)
```

Sure, we could do that. My point about keeping a single array with optional subkeys was:

- Dynamic dependencies are less common than static ones, so you wouldn't need to use subkeys that often.
- It's closer to `wp_register_script`, which is something people are familiar with.
- If someone adds a dynamic dependency as a static one (forgets using subkeys), it's not optimal, but it's not broken. Also, they will see a warning in the devtools console.

youknowriad replied:
what about the point about making both experiments compatible in terms of APIs?

luisherranz replied:
What do you mean? So we can hypothetically move from option 3 (required dependencies) to option 2 (optional dependencies) in the future?

youknowriad replied:
yes, maybe it's not that important thought and we should just make the right decision :)

luisherranz replied:
By the way, there's an alternative API that I don't dislike because it opens the door for other dependency properties in the future:

- Single dependencies array:
  ```php
  wp_register_module( 'my-module', '/src', array( 'static-dep' ) );
  ```
- But dependencies can be strings or arrays if they have properties:
  ```php
  wp_register_module(
    'my-module',
    '/src',
    array(
      'static-dep',
      array( 'id' => 'dynamic-dep', 'type' => 'dynamic' ),
    )
  );
  ```
- Other properties could be added if necessary in the future, for example, if we'd ever want to support `scopes`:
  ```php
  wp_register_module(
    'my-module',
    '/src',
    array(
      array( 'id' => 'static-dep', 'version' => '^3.0.0' ),
    )
  );
  ```
- This one also prioritizes the simplicity of the most common case, which is static dependencies.

luisherranz replied:
We are not using any dynamic dependency yet, so I'll make this change in a subsequent PR.

luisherranz committed:
Add polyfill

luisherranz committed:
Fix $version using its own arg in register calls

cbravobernal committed:
Add unit tests

cbravobernal committed:
Refactor tests, add test get import map

cbravobernal commented on lib/experimental/modules/class-gutenberg-modules.php:
```diff
@@ -0,0 +1,186 @@
+		foreach ( self::$enqueued as $module_identifier ) {
+			if ( isset( self::$registered[ $module_identifier ] ) ) {
+					$module = self::$registered[ $module_identifier ];
+					echo '<script type="module" src="' . $module['src'] . self::get_version_query_string( $module['version'] ) . '" id="' . $module_identifier . '"></script>';
```
Should we add here a integrity check for array keys 'src' and 'version' ?

gziolo replied:
There are `wp_print_script_tag` and `wp_get_script_tag` that are filterable and they provide basic sanitization for the values provided in attributes.

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/56143/commits/4e6c5da06fd85662b1c6eb120e382af455958200. Thanks folks!

cbravobernal committed:
Cleaning data

cbravobernal commented on lib/experimental/modules/class-gutenberg-modules.php:
```diff
@@ -0,0 +1,186 @@
+	/**
+	 * Returns the import map array.
+	 *
+	 * @return string The import map.
```
This is returning an array, not a string.

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/56143/commits/829c6ec177c3d03eb5d801199bf99b8b9d455889. Thanks, Carlos!

gziolo commented on lib/experimental/interactivity-api/modules.php:
```diff
@@ -0,0 +1,33 @@
+function gutenberg_register_interactivity_module() {
+	gutenberg_register_module(
+		'@wordpress/interactivity',
+		'/wp-content/plugins/gutenberg/build/interactivity/index.min.js',
```
Here and in other places to make the path work with different WP configurations:
```suggestion
		gutenberg_url( 'build/interactivity/index.min.js' ),
```

luisherranz replied:
Fixed in https://github.com/WordPress/gutenberg/pull/56143/commits/5488eab4d7021e6dfcb20f0e9685555b8ef09016. Thanks, Greg.

luisherranz committed:
Use gutenberg_url()

luisherranz committed:
Use wp_print_script_tag

luisherranz committed:
Fix DocBlock on get_import_map()

luisherranz committed:
Merge branch 'trunk' into experiment/modules-and-import-maps-api-required-server-dependency-graph

luisherranz committed:
Load navigation module only on Gutenberg plugin

luisherranz committed:
Load query module only on Gutenberg plugin

luisherranz committed:
Load search module only on Gutenberg plugin

luisherranz committed:
Load file module only on Gutenberg plugin

luisherranz committed:
Load image module only on Gutenberg plugin

luisherranz committed:
Make registration optional

luisherranz committed:
Improve navigation logic

luisherranz commented on lib/compat/wordpress-6.5/class-wp-navigation-block-renderer.php:
```diff
@@ -547,20 +547,28 @@ private static function get_nav_element_directives( $should_load_view_script ) {
 	 */
 	private static function handle_view_script_loading( $attributes, $block, $inner_blocks ) {
 		$should_load_view_script = static::should_load_view_script( $attributes, $inner_blocks );
+		$is_gutenberg_plugin     = defined( 'IS_GUTENBERG_PLUGIN' ) && IS_GUTENBERG_PLUGIN;
```
Not sure if this is needed because it's outside of the `block-library` folder.

gziolo replied:
As far as I know, this class needs to be backported to WordPress core. I saw @youknowriad and @tellthemachines discussing options, including moving it to the block folder to automatically sync it from npm to WP core. In that case, the check would be necessary.

luisherranz replied:
Let's keep it for now, then. Thanks, Greg.

tellthemachines replied:
Might be good to keep this in even if the file isn't auto-synced so that it's clear which parts of the logic should be added to core. 

Is this switch to modules planned for 6.5, or will it stay experimental for a while yet?

luisherranz replied:
We don't know yet. But if the code is ready and there's consensus on the API, it could be 🙂

luisherranz commented on packages/block-library/src/search/index.php:
```diff
@@ -203,27 +213,17 @@ function register_block_core_search() {
- *
- * @global WP_Scripts $wp_scripts
- */
-function block_core_search_ensure_interactivity_dependency() {
```
To be honest, I don't know why we had this function in the first place as the view script of the search block always has the `wp-interactivity` dependency.

@SantosGuillamot do you remember?

SantosGuillamot replied:
I don't remember either, to be honest. It seems it was included in this pull request in case it helps: [link](https://github.com/WordPress/gutenberg/pull/54594).

luisherranz replied:
Ok, I don't think it's needed then. Thanks, Mario 🙂

luisherranz committed:
Remove unnecessary check

luisherranz committed:
Fix missing view file

luisherranz committed:
Don't print the import map if it's empty

luisherranz committed:
Load the importmap polyfill locally

luisherranz committed:
Move the es-modules-shims package to the top

luisherranz committed:
Use the public functions in the tests

luisherranz reviewed (commented):
Ok, everything is working great here, so even though this started as an experiment, I think we can land it to test it out in real sites.

_Kudos to @c4rl0sbr4v0 for adding the unit tests._

gziolo commented on phpunit/experimental/modules/class-gutenberg-modules-test.php:
```diff
@@ -0,0 +1,125 @@
+		gutenberg_register_module( $module_identifier, $src, $dependencies, $version );
+
+		// Use reflection to access the private static property $registered.
+		$reflection = new ReflectionClass( 'Gutenberg_Modules' );
```
The unit tests included focus on testing internal implementation which will make it harder to refactor code later. How much code would you be able to test without using reflections but through public API?

cbravobernal replied:
We could move to e2e and test 
 `get_import_map` which uses `get_dependencies` and `enqueue`
`print_enqueued_modules`, which uses `get_version_query_string` and needs `register`

With only those two tests we would be covering the entire class. Let me update it.

gziolo replied:
For reference, this is how regular scripts are covered with unit tests in WordPress scripts:

https://github.com/WordPress/wordpress-develop/blob/96339bda237182d2d1246b8b4f12df9833643cd9/tests/phpunit/tests/dependencies/scripts.php

These tests cover in many places `wp_print_scripts` and use a custom assertion `assertEqualMarkup` to compare the expected output with the result.

cbravobernal replied:
I'm afraid we don't have the `assertEqualMarkup` and `parse_markup_fragment` functions in Gutenberg. 

I had to use `WP_HTML_Tag_Processor` instead, just making sure the `src` and the `type` of the scripts enqueued are correct.

I can migrate it to those structure on the backport once is tested in Gutenberg.

luisherranz replied:
I like the usage of the `WP_HTML_Tag_Processor` 🙂

youknowriad commented on lib/experimental/interactivity-api/modules.php:
```diff
@@ -0,0 +1,33 @@
+ * Register the `@wordpress/interactivity` module.
+ */
+function gutenberg_register_interactivity_module() {
+	gutenberg_register_module(
```
We should have a way in our code to have an array of "esm modules" and have these be registered automatically (similarly to the other types of packages we have: non modules)

youknowriad commented on packages/block-library/src/file/index.php:
```diff
@@ -15,19 +15,29 @@
+				$block->block_type->view_script_handles = array_merge( $script_handles, array( $view_js_file ) );
+			}
 		}
 	}
```
This feels like complex logic, do we expect any interactive block to write this kind of code? Is there a way to streamline that?

gziolo replied:
Similar feedback before WP 6.4 Beta 1: https://github.com/WordPress/wordpress-develop/pull/5262#discussion_r1336258326.

Follow-up ticket in Trac: https://core.trac.wordpress.org/ticket/59462.

youknowriad commented on packages/block-library/src/image/index.php:
```diff
@@ -362,5 +366,14 @@ function register_block_core_image() {
+			array( '@wordpress/interactivity' ),
+			defined( 'GUTENBERG_VERSION' ) ? GUTENBERG_VERSION : get_bloginfo( 'version' )
+		);
+	}
```
Same about these calls, we should have a way to register these automatically based on block.json interactivity key or something. We shouldn't be repeating these over and over again.

luisherranz commented:
@youknowriad, about `block.json` registration of interactive blocks: 

Today, the `viewScript` field assumes a 1:1 relation (this blocks <-> this script/s), which is not a great fit when blocks are conditionally interactive. This turned out to be more common than we initially thought, to the point that all currently interactive core blocks are conditionally interactive. 

That weird `$block->block_type->view_script_handles` logic is there to keep the scripts in the `viewScript` fields, but remove or add them conditionally. I agree it's not ideal, but improving it is not a blocker for the Interactivity API, so we haven't been focusing on that.

If we want to keep the script/module declarations in the `block.json` (which I think is great), we need to rethink that 1:1 relation, among a few other things that interconnect, like behaviors (the possibility of adding scripts/modules to a block by a third-party) and [lazy hydration](https://github.com/WordPress/gutenberg/discussions/52723).

On the other hand, `wp_register_script/module` doesn't have that 1:1 relation because they can be called conditionally in the code. And the reason the first versions of these experiments were much simpler was because I replaced the `viewScript` fields with a `gutenberg_register_module` call:

<img width="800" alt="Screenshot 2023-11-29 at 09 44 11" src="https://github.com/WordPress/gutenberg/assets/3305402/5c424022-1a1b-4e76-86e7-286cf4739760">

\
But I brought back the weird `$block->block_type->view_script_handles` logic in this experiment because now that we want to land this in Gutenberg, @gziolo reminded me that all the code in `block-library` should be ready to be synced with WordPress at any moment (not just before the beta releases). So, I added the `viewScript` fields and the `$block->block_type->view_script_handles` logic again behind all those `$is_gutenberg_plugin` checks to make sure that WordPress Core can still use those blocks without `gutenberg_register_module`.

Finally, this is my opinion about this matter: 
- I absolutely agree there is potential to keep absorbing the block asset declaration in the block metadata, and I think it's a good direction to improve the developer experience, but due to the conditional nature of the frontend interactivity, it needs careful thinking. I do have some vague ideas in mind, but nothing concrete or useful yet.
- It's not a blocker to release the 1.0 version of the Interactivity API, so we haven't been focusing on that. It's an improvement that can be done later.
- It's not a blocker to release a minimal version of a Modules API either, and whatever we build, we will build it on top of the PHP APIs anyway, so this is the first step to me.

So, I won't wait until we have those improvements to merge this experiment.

youknowriad commented:
> I absolutely agree there is potential to keep absorbing the block asset declaration in the block metadata, and I think it's a good direction to improve the developer experience, but due to the conditional nature of the frontend interactivity, it needs careful thinking. I do have some vague ideas in mind, but nothing concrete or useful yet.

I think it's fine for enqueuing to be conditional, I'm thinking that registration should be more automatic though: the following is fine in terms of devx.

```
if ( $shouldBeInteractive ) {
    wp_enqueue_module( 'some-module' ) 
}
```

---
Not saying any of this is a blocker but the potential for error now is too big. Unless you know what you're doing (like you do), I'm sure developers will mess things up. I want clear answers to these questions:

 - How do I write a new ESmodule package in Gutenberg Core. Which arrays/config should I change to define that this new package is ESmodule.
 - How do I use my new ESModule. Does `import( '@wordpress/my-package' )` work? (See my example esm PR)

Now separately, for interactive blocks, I expect similar answers (docs)

 - How do I write a new interactive block in Gutenberg.
 - How do I add conditional interactivity.

Let's remove the "guess" work from developers (first core developers), later third-party when we make the API public.

luisherranz commented:
I think you are assuming that we are neglecting all those questions, but we are not.

I do agree that we need to work on that, but again, the low-level Modules API and how to leverage that within the block metadata (or how to use Webpack/Esbuild to generate modules) are separate things.

As I said in the [Tracking Issue](https://github.com/WordPress/gutenberg/issues/55942), the goal of this initiative is to work on the low-level primitives only:

> The goal of this initiative is to implement the necessary low-level primitives to add native support in WordPress for registering and enqueueing JavaScript modules, including generating an import map.

I think we can land this to start shipping modules to real sites and keep iterating the low-level API. Maybe try it out in Jarda's Editor experiments as well. Once the API is clearer, we can start thinking about the `block.json` integration and other optimizations.

I reiterate my thoughts that figuring out how to integrate the Modules API with the block metadata is not a blocker for either the Modules API or the 1.0 version of the Interactivity API, so even if devs need to use this syntax and a custom Wepabck/Esbuild config to be able to use interactive blocks for 1 or 2 WordPress versions, it'd be fine to me:

```php
function register_my_plugin() {
	register_block_type_from_metadata(
		__DIR__ . '/file',
		array( 'render_callback' => 'render_my_plugin' )
	);

	wp_register_module(
		'@my-plugin/some-block',
		'...',
		array( '@wordpress/interactivity' ),
		'1.2.3'
	);
}
add_action( 'init', 'register_my_plugin' );

function render_my_plugin( $attributes ) {
	if ( $attributes['someInteractiveAttribute'] ) {
		wp_enqueue_module( '@my-plugin/some-block' );
	}
}
```

That said, if someone else has the time and willingness to start experimenting with the `wp-scripts` and `block.json` integration now, please don't hesitate to do so 🙂

youknowriad commented:
> I think you are assuming that we are neglecting all those questions, but we are not.

Not really :) I'm just stating what's important for me personally. I largely prefer higher level APIs over lower level ones because we can control/optimize... and avoid bugs. I often find that writing docs (even for internal contributors) help us shape the right APIs and answer the right questions.

luisherranz commented:
Haha, ok. Thanks, Riad 🙂

So, what are your suggested next steps then?

youknowriad commented:
I'm fine with merging this one personally. But would love some follow-up documentation to answer the questions here.

> How do I write a new ESmodule package in Gutenberg Core. Which arrays/config should I change to define that this new package is ESmodule.
> How do I use my new ESModule. Does import( '@wordpress/my-package' ) work? (See my example esm PR)

And the interactivity helpers/docs could be explored as well.

youknowriad commented:
I've been thinking more here. This modules PR is centered towards Interactivity API.

I'd like personally to understand the story of using ESmodules in our current code. Let's say I want to introduce a new package called `@wordpress/utils` and I want that to be a module.

Can I just do the following in `edit-post` for instance to have everything working properly?

```js 
import { myUtil } from '@wordpress/utils';
```

I mean how can existing scripts make use of modules. Would that automatically add the package to the import map? Would that load the module synchronously or will I be forced to always to do the following If I want to use a module in a non-module?

```js
const { myUtil } = await import( '@wordpress/utils' );
```

this is fine too, but just want some clarity I guess.

cbravobernal commented on lib/experimental/modules/class-gutenberg-modules.php:
```diff
@@ -0,0 +1,195 @@
+	 */
+	public static function print_module_preloads() {
+		foreach ( self::get_dependencies( self::$enqueued, array( 'static' ) ) as $dependency_identifier => $module ) {
+				echo '<link rel="modulepreload" href="' . $module['src'] . self::get_version_query_string( $module['version'] ) . '" id="' . $dependency_identifier . '">';
```
This id would be duplicated with the `print_enqueued_modules`. It would end up printing:

`<link rel="modulepreload" href="module_1_src?ver=1.0.0" id="module_1">`
`<script type="module" src="module_1_src?ver=1.0.0" id="module_1">`

is that ok?

luisherranz replied:
Good catch. The preload is unnecessary in that case, so we have to filter the enqueued modules out of the preloaded ones.

luisherranz replied:
I've added it to the follow-up tasks in the PR description. I'll add it to the Tracking Issue as well once/if this is merged.

luisherranz commented:
> how can existing scripts make use of modules

Let's discuss that in the Tracking Issue.

cbravobernal committed:
Update test to be more output oriented

cbravobernal committed:
Remove we from comments.

luisherranz committed:
Merge branch 'trunk' into experiment/modules-and-import-maps-api-required-server-dependency-graph

luisherranz committed:
Start using modules in the interactivity e2e tests

luisherranz commented on phpunit/experimental/modules/class-gutenberg-modules-test.php:
```diff
@@ -0,0 +1,66 @@
+		$tags = new WP_HTML_Tag_Processor( $import_map_markup );
+		$this->assertEquals( true, $tags->next_tag( array( 'rel' => 'importmap' ) ) );
+
+		// Test that there is 1 <link type="modulepreload"> added to the markup.
+		$tags = new WP_HTML_Tag_Processor( $preload_markup );
+		$this->assertEquals( true, $tags->next_tag( array( 'rel' => 'modulepreload' ) ) );
```
This should fail once we filter the enqueued modules on the preloads, but I'll adjust that on the PR that fixes that, and I'll add more tests to cover both cases.

luisherranz committed:
Update package-lock.json

cbravobernal reviewed (approved):
> I'm fine with merging this one personally.

I'll take that as a second approval. I've tested and works as expected.
All the concerns have been discussed or updated, and all the code is covered by tests.

briangrider commented:
Hi all, I'm a third party dev who has been experimenting with the Interactivity API the past few versions of Gutenberg and notice that with this RC, the API is no longer working for third party blocks (at least the way it used to with just  adding a viewScript and the data-wp-interactive attribute to interactive blocks). I'm getting the following errors with blocks/code that was working previously:

VM3154 index.min.js:1 Uncaught SyntaxError: Unexpected token 'export' (at VM3154 index.min.js:1:33535)
external window ["wp","interactivity"]:1 Uncaught TypeError: Cannot read properties of undefined (reading 'interactivity')
    at @wordpress/interactivity (external window ["wp","interactivity"]:1:1)
    at __webpack_require__ (bootstrap:19:1)
    at make namespace object:7:1
    at view.js?ver=99a0ba6c6dba75ed65c4:179:3
    at view.js?ver=99a0ba6c6dba75ed65c4:181:12
@wordpress/interactivity @ external window ["wp","interactivity"]:1
__webpack_require__ @ bootstrap:19
(anonymous) @ make namespace object:7
(anonymous) @ view.js?ver=99a0ba6c6dba75ed65c4:179
(anonymous) @ view.js?ver=99a0ba6c6dba75ed65c4:181
thing/:169

Can someone point me to the new method for registering/using interactive third party blocks? Please let me know if there is a better place to ask this question. Thanks!

luisherranz commented:
I'm working on a migration guide to explain how to use modules. I'll let you know once it's ready.

briangrider commented:
Thanks @luisherranz, greatly appreciated. 🙏 Moving back to 17.02 for the time being.

youknowriad commented:
Looks like this PR had a small but noticeable impact on LCP - TTFB metrics https://www.codevitals.run/project/gutenberg

luisherranz commented:
In the Editor? That doesn't make sense…

youknowriad commented:
@luisherranz no in the frontend.

luisherranz commented:
That would imply that modules are slower than deferred scripts, which should not be.

What is it testing? Which site/theme/blocks?

luisherranz commented:
> That would imply that modules are slower than deferred scripts, which should not be.

Oh, there is a difference indeed, the polyfill!

Let me make a PR to load it only when there's an import map and only when the browser doesn't support it, to see if the numbers improve.

youknowriad commented:
@oandregal knows more about this test but I believe we're used a fixed version of "twentytwentythree“ with the default content.

luisherranz commented:
Is there a way to see if this PR improves those numbers?

- https://github.com/WordPress/gutenberg/pull/56699

youknowriad commented:
Yes, you can check the logs of the performance job https://github.com/WordPress/gutenberg/actions/runs/7051192339/job/19193537419?pr=56699 

But there's a margin of error, if the change is small, it's better to confirm it in the graphs post merge.

luisherranz commented:
It didn't improve the metrics, so I'll do more tests. I'll report my findings in [the new PR](https://github.com/WordPress/gutenberg/pull/56699).

briangrider commented:
@luisherranz I was able to figure out how to get this all working with the module api/interactivity. Very cool features and I can see how huge this is for WP. What's the best way to stop webpack from trying to bundle the @wordpress/interactivity import so view.js files can still be built without breaking on the front end?

briangrider commented:
@luisherranz Yep, this is what I was doing too and it works fine! Like you said, not having multiple files and no minification is where I was getting hung up but it doesn't seem like it's too big of a deal at this point because a standard viewScript (at least the ones I'm writing) tends to be around 100-200 lines or less. I think once this is in core, it'll be important to be able to use the standard wp-scripts package but in the interim, this is an easy enough way to be able to experiment with this awesome new functionality

t-hamano commented:
Loading modules does not seem to work with classic themes, as reported by #57370. In the classic theme, only polyfill JS is loaded.

Block Theme:

```html
<!DOCTYPE html>
<html lang="ja">
<head>
	<!-- ... -->
	<script type="importmap">
	{"imports":{"@wordpress\/interactivity":"http:\/\/localhost:8888\/wp-content\/plugins\/gutenberg\/build\/interactivity\/index.min.js?ver=6.4.2"}}
	</script>
	<script type="module" src="http://localhost:8888/wp-content/plugins/gutenberg/build/interactivity/image.min.js?ver=6.4.2" id="@wordpress/block-library/image"></script>
	<script type="module" src="http://localhost:8888/wp-content/plugins/gutenberg/build/interactivity/navigation.min.js?ver=6.4.2" id="@wordpress/block-library/navigation-block"></script>
	<!-- ... -->
</head>
<body>
	<!-- ... -->
	<script src="http://localhost:8888/wp-content/plugins/gutenberg/build/modules/importmap-polyfill.min.js" defer=""></script>
</body>
</html>
```

Classic Theme:

```html
<!DOCTYPE html>
<html lang="ja">
<head>
</head>
<body>
	<!-- ... -->
	<script src="http://localhost:8888/wp-content/plugins/gutenberg/build/modules/importmap-polyfill.min.js" defer=""></script>
</body>
</html>
```

My guess is that the reason why it doesn't work is as follows:

Each block's module JS is enqueued within the block's rendering callback function ([Image block example](https://github.com/WordPress/gutenberg/blob/0dd6255ffa8c1d3261d861d4b589afc3c17528fa/packages/block-library/src/image/index.php#L55)). Based on the enqueued JS, it outputs the import map and module script via [the wp_head hook](https://github.com/WordPress/gutenberg/blob/0dd6255ffa8c1d3261d861d4b589afc3c17528fa/lib/experimental/modules/class-gutenberg-modules.php#L265C1-L272).

For block themes, the entire HTML is [pre-fetched](https://github.com/WordPress/wordpress-develop/blob/0e284447951869fd3ff58b345ec070b0ab97d9bc/src/wp-includes/template-canvas.php#L12) and the entire page is displayed. Here, the block's rendering callback is called and the block's module JS is enqueued. This is before the `wp_head` action hook is executed.

But in the classic theme, the block is rendered when the content is rendered, i.e. when `the_content` is executed within the page. When the `wp_head` action hook is executed, the blocks have not yet been rendered, meaning the module JS for each block has not been enqueued, so the import map and module JS will not be output.

I don't have any good ideas at the moment, but I think some logic adjustments are needed.

luisherranz commented:
Thanks for the report @t-hamano, and thanks for the fix, @c4rl0sbr4v0:

- https://github.com/WordPress/gutenberg/pull/57396
